### PR TITLE
Add model regularization via label smoothing.

### PR DESCRIPTION
**Pull request by helson73**
Label smoothing. 
Reference:
 https://arxiv.org/abs/1512.00567

### CODE REVIEW

**Comment by helson73** [comment]
Training option: 
`python -u train.py -gpuid 0 -save_model models/100k/ja-en -data data/ja-en-100k -layers 4 -rnn_size 512 -word_vec_size 512 -epochs 20 -max_grad_norm 0 -optim adam -encoder_type transformer -decoder_type transformer -position_encoding -dropout 0.2 -param_init 0 -warmup_steps 2000 -learning_rate 0.05 -decay_method noam -label_smoothing 0.1`

Before:
Train perplexity: 32.4537
Train accuracy: 38.7784
Validation perplexity: 56.2081
**Validation accuracy: 33.0021**

After PR #456:
Train perplexity: 29.8781
Train accuracy: 40.5064
Validation perplexity: 52.0337
**Validation accuracy: 34.0583**

After PR #462(this):
Train perplexity: 79.9049
Train accuracy: 40.6099
Validation perplexity: 128.389
**Validation accuracy: 34.5876**

As "Attention is All You Need" reported, accuracy is slightly improved. I will do further test on larger dataset with longer epochs.

**Comment by srush** [comment]
Fantastic this is such a clean implementation. 

Just to check though, should PPL be that much worse? I know it should get worse, but this seems like a lot.

**Comment by helson73** [comment]
@srush 
Thanks, PPL looks weird because technically it's exp(h(q, p)) rather than exp(nll(p)). Exact PPL would be smaller.

**Comment by srush** [comment]
Wait, I don't understand. Isn't that the whole point of subtracting out the the additional term?

**Comment by srush** [comment]
Sorry, I was being unclear. I agree it is calculating `exp(h(q, p))`, but maybe we could switch to `exp(nll(p))` for validation, so that we can know the fair comparison? 

It is a little confusing that we call it perplexity, when it really isn't. 

Also you should join our gitter channel. https://gitter.im/OpenNMT/OpenNMT-py easier to discuss.
